job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+
+      var contextName = ""
+      do {
+        contextName = java.util.UUID.randomUUID().toString().substring(0, 8) + "-" + classPath
+      } while (contexts contains contextName)
+
+      startContext(contextName, mergedConfig, true) { ref =>
+        originator ! (contexts(contextName), resultActors(contextName))
+      } { err =>
+        originator ! ContextInitError(err)
+      }
+
+    case GetResultActor(name) =>
+      sender ! resultActors.get(name).getOrElse(globalResultActor)
+
+    case GetContext(name) =>
+      if (contexts contains name) {
+        sender ! (contexts(name), resultActors(name))
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case StopContext(name) =>
+      if (contexts contains name) {
+        logger.info("Shutting down context {}", name)
+
+        context.watch(contexts(name))
+        contexts(name) ! PoisonPill
+        resultActors.remove(name)
+        sender ! ContextStopped
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case Terminated(actorRef) =>
+      val name: String = actorRef.path.name
+      logger.info("Actor terminated: {}", name)
+      contexts.foreach { kv => if (kv._2 == actorRef) contexts.remove(kv._1) }
+      resultActors.foreach { kv => if (kv._2 == actorRef) resultActors.remove(kv._1) }
+  }
+
+  private def initContext(actorName: String, ref: ActorRef, timeoutSecs: Long = 1)
+                         (ctxName: String, ctxConf: Config, isAdHoc: Boolean)
+                         (successFunc: ActorRef => Unit, failureFunc: Throwable => Unit): Unit = {
+    import akka.pattern.ask
+
+    val resultActor = if (isAdHoc) globalResultActor else context.actorOf(Props(classOf[JobResultActor]))
+    (ref ? JobManagerActor.Initialize(
+      daoActor, Some(resultActor), ctxName, ctxConf, isAdHoc, self))(Timeout(timeoutSecs.second)).onComplete {
+      case Failure(e:Exception) =>
+        logger.info("Failed to send initialize message to context " + ref, e)
+        ref ! PoisonPill
+        failureFunc(e)
+      case Success(JobManagerActor.InitError(t)) =>
+        logger.info("Failed to initialize context " + ref, t)
+        ref ! PoisonPill
+        failureFunc(t)
+      case Success(JobManagerActor.Initialized(_)) =>
+        logger.info("SparkContext {} joined", ctxName)
+        contexts(ctxName) = ref
+        resultActors(ctxName) = resultActor
+        successFunc(ref)
+    }
+
+  }
+
+  private def startContext(name: String, contextConfig: Config, isAdHoc: Boolean)
+                          (successFunc: ActorRef => Unit)(failureFunc: Throwable => Unit) = {
+    require(!(contexts contains name), "There is already a context named " + name)
+    val contextActorName = "jobManager-" + java.util.UUID.randomUUID().toString.substring(16)
+
+    logger.info("Starting context with actor name {} ", contextActorName)
+
+    val pb = Process(s"$managerStartCommand $contextActorName ${selfAddress.toString}")
```

Comment:
@Ankit1010 I thought about this some more.  The code is like this:
```scala
        val memberActors = RootActorPath(member.address) / "user" / "*"
        context.actorSelection(memberActors) ! Identify(memberActors)
```

Thus right now actor identity is being invoked on all actors of the forked process that reports MemberUp.  This would include the JobManagerActor, JarManager, JobStatusActor, and any other actor that gets spun up.   Even if we used the subtree idea, any actors spun up by the JobManager itself would appear under the subtree.  Thus, I do believe we need not just the pure context name as the actor name, but something else to differentiate the name.  As for multiple master, we have to sketch that out much further. :)